### PR TITLE
begin the repo linting 

### PR DESCRIPTION
**Pull request by jGaboardi**
This PR kicks off #589. Should we do this:
1. piece-meal one small PR at a time; or
2. do one PR, but go incrementally with small chunks + reviews

(This process is going to be a slog... between 2k-3k fixes on first inspection)

### CODE REVIEW

**Comment by martinfleis** [comment]
I would do one PR that contains only the automatic fixes by ruff but across the whole code base and than a series of smaller PRs with the rest. The former does not need a ton of review and may resolve quite a lot already.

File: pyproject.toml [hunk 1]
```diff
@@ -93,14 +93,18 @@ ignore = [
...
     "N803",
     "N806",
     "N999",
     "UP007"
 ]
 exclude = ["libpysal/tests/*", "docs/*"]
+[tool.ruff.per-file-ignores]
+"*__init__.py" = [
+    "F401",  # imported but unused
+    "F403",  # star import; unable to detect undefined names
```
Comment (martinfleis) [inline pyproject.toml:105]: I'd like to eliminate star imports if possible but that can be a follow-up at some point in the future. It just may be worth opening an issue about it?

**Comment by martinfleis** [comment]
That said, I am happy to merge this as is.

**Comment by jGaboardi** [comment]
> I would do one PR that contains only the automatic fixes by ruff but across the whole code base and than a series of smaller PRs with the rest. The former does not need a ton of review and may resolve quite a lot already.

The problem with autofixing everything is that is sometimes breaks things. For example, in [`libpysal/weights/__init__.py`](https://github.com/pysal/libpysal/pull/597/files#diff-1639c0182a1375d17f3d4676210729c0d8e2fd15c013b355fc2e5e7f83902efdL1) the `from .weights import * # noqa I001` import location can't change or we see the following:

```
ImportError while importing test module '/Users/userlibpysal/libpysal/weights/tests/test_weights_IO.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../mambaforge/envs/py311_libpysal/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
libpysal/__init__.py:27: in <module>
    from . import cg, examples, io, weights
libpysal/examples/__init__.py:14: in <module>
    from .base import example_manager
libpysal/examples/base.py:17: in <module>
    from ..io import open as ps_open
libpysal/io/__init__.py:2: in <module>
    from .iohandlers import *
libpysal/io/iohandlers/__init__.py:6: in <module>
    from . import dat
libpysal/io/iohandlers/dat.py:2: in <module>
    from ...weights import W
libpysal/weights/__init__.py:3: in <module>
    from .gabriel import Delaunay, Gabriel, Relative_Neighborhood
libpysal/weights/gabriel.py:3: in <module>
    from libpysal.weights import W, WSP
E   ImportError: cannot import name 'W' from partially initialized module 'libpysal.weights' (most likely due to a circular import) (/Users/user/libpysal/libpysal/weights/__init__.py)
```

File: pyproject.toml (hunk 1 above)
Comment (jGaboardi) [reply pyproject.toml:105]: I completely agree about star imports; and making a separate issue.

**Comment by martinfleis** [comment]
> The problem with autofixing everything is that is sometimes breaks things

fun ... okay, let's go piece by piece then

**Review by martinfleis — APPROVED** [review]

**Pull request merged**